### DISCUSSION TITLE
Server communicating SSL_accept error back to client during SSL_connect

### DISCUSSION THREAD

**Discussion by TheVeryOldMan** [openssl/openssl#25656] [Q&A]
Basic setup:
1. client, server connect via regular socket.
2. each create CTX/SSL.
3. Both SSL_set_fd to their original socket.
4. Server sets SSL_CTX_set_verify() to peer verify & fail if no peer cert sent, then goes into SSL_accept.
5. Client calls SSL_connect without setting up a certificate.
6. Server SSL_accept() fails as expected and gets error message.
7. Client SSL_connect returns success even though handshake failed.

Now the problem:
1. Client sees SSL_connect ok, eventually comes around to SSL_write and gets socket error, no surprise, but the client program does not know why.
2. Somehow I need to get this error message "SSL error: peer did not return a certificate" from the server back to the client.
3. Since the SSL connection is not up and I cannot SSL_write() from the server, I figured I could use the original client/server socket to send it back.
4. This "sort of" works however there are 24 bytes in front followed by the message (maybe some goop that SSL_free() wrote?).
5. IOW something is still sitting on top of the original sockets. The sent/received byte counts match.

The solution I need to figure out is 
1. Client is notified that SSL handshake failed
2. Client gets error message

Thanks,
Paul

**Comment by mattcaswell** [comment]
In this scenario the server should send a certificate_required alert to the client.

I can simulate the situation you are describing using OpenSSL's s_server and s_client tools. If I start s_server like this:

````
$ openssl s_server -Verify 5 -cert cert.pem -key key.pem
````

Then connect to it using s_client:

````
$ openssl s_client
````

Then you should see this error as the last line of output from s_client:
````
80CB4FF637770000:error:0A00045C:SSL routines:ssl3_read_bytes:tlsv13 alert certificate required:ssl/record/rec_layer_s3.c:910:SSL alert number 116
````
Basically when s_client tries to read data from the server (via `SSL_read`) instead of getting the application data it expected, it instead reads this alert and fails with the above error on the OpenSSL error stack.

You should be able to do the same thing in your code.

**Reply by TheVeryOldMan** [reply]
Thank you for digging through my long message, I will do this and see if I can figure out what calls openssl is making vs what I am doing. My openssl-based client works with my Windows SChannel server however looking closely at my SChannel code I see that the failed ssl connection is not closed until after the message is sent, iow somehow a failed connection is still able to send data. Again thank you for the direction.

**Reply by TheVeryOldMan** [reply]
I compiled openssl with -g, gdb single stepped my way through it, and finally have it working!

First you must SSL_clear_mode SSL_MODE_AUTO_RETRY. If not, then your read will hang.

Then after returning from SSL_connect with 1, set up a select readfds, then SSL_read(sslp,buf,1024).

If SSL_read returns 0, then SSL_get_error(), SSL_free(sslp), and display the error, e.g. "tlsv13 alert certificate required".

If it's (-1), then SSL_get_error() should return SSL_ERROR_WANT_READ.

Go back to your set up the select, SSL_read, SQL_get_error() should return SSL_ERROR_WANT_READ again and you are ready to go. BTW you can set SSL_MODE_AUTH_RETRY back on when this is done.
